Sources/Build/Destination.swift

```diff
@@ -0,0 +1,128 @@
+import Foundation
+import Basic
+import PackageModel
+
+enum DestinationError: Swift.Error {
+    /// Couldn't find the Xcode installation.
+    case invalidInstallation(problem: String)
+
+    /// Wrong version.
+    case destinationFileVersionMismatch(problem: String)
+}
+
+extension DestinationError: CustomStringConvertible {
+    var description: String {
+        switch self {
+        case .destinationFileVersionMismatch(let problem):
+            return problem
+        case .invalidInstallation(let problem):
+            return problem
+        }
+    }
+}
+
+/// The compilation destination, has information about everything that's required for a certain destination.
+public struct Destination {
+
+    /// The clang/LLVM triple describing the target OS and architecture.
+    public let target: String
+
+    /// The SDK used to compile for the destination.
+    public let sdk: AbsolutePath
+
+    /// The toolchain containing the compilers/linker to be used for the compilation.
+    public let toolchain: AbsolutePath
+
+    /// The file extension for dynamic libraries (eg. `.so` or `.dylib`)
+    public let dynamicLibraryExtension: String
+
+    /// Additional flags to be passed to the C compiler.
+    public let extraCCFlags: [String]
+
+    /// Additional flags to be passed to the Swift compiler.
+    public let extraSwiftCFlags: [String]
+
+    /// Additional flags to be passed when compiling with C++.
+    public let extraCPPFlags: [String]
+
+    /// The destination describing the host OS.
+    public static func hostDestination() throws -> Destination {
+        #if os(macOS)
+            let sdkPath = try Process.checkNonZeroExit(args: "xcrun", "--sdk", "macosx", "--show-sdk-path").chomp()
+            guard !sdkPath.isEmpty else {
+                throw DestinationError.invalidInstallation(problem: "could not find default SDK")
+            }
+
+            let platformPath = try Process.checkNonZeroExit(args: "xcrun", "--sdk", sdkPath, "--show-sdk-platform-path").chomp()
+            guard !platformPath.isEmpty else {
+                throw DestinationError.invalidInstallation(problem: "could not find SDK platform path for SDK \(sdkPath)")
+            }
+
+            let sdkPlatformFrameworksPath = AbsolutePath(platformPath).appending(components: "Developer", "Library", "Frameworks")
+
+            let devDirPath = try Process.checkNonZeroExit(args: "xcode-select", "-p").chomp()
+            let platformFrameworkPath = AbsolutePath(sdkPath).appending(component: "System")
+                                                             .appending(component: "Library")
+                                                             .appending(component: "Frameworks").asString
+            let toolchainPath = AbsolutePath(devDirPath).appending(component: "Toolchains")
+                                                        .appending(component: "XcodeDefault.xctoolchain")
```

Comment:
This isn't going to work with custom toolchains. I think we should store binDir instead of toolchain, then compute the binDir.